src/ctelnet.cpp

```diff
@@ -1519,9 +1672,133 @@ void cTelnet::processTelnetCommand(const std::string& telnetCommand)
     case TN_SB: {
         option = telnetCommand[2];
 
+        // NEW_ENVIRON
+        if (option == OPT_NEW_ENVIRON && enableMNES) {
+            QByteArray payload = QByteArray::fromRawData(telnetCommand.c_str(), telnetCommand.size());
+
+            if (telnetCommand.size() < 7) {
+                return;
+            }
+
+            // Trim off the Telnet suboption bytes from beginning (3) and end (2)
+            payload = payload.mid(3, static_cast<int>(payload.size()) - 5);
+
+            if (payload.data()[0] == MNES_SEND && payload.data()[1] == MNES_VAR) {
+                const QMap<QString, QString> environVariables = getEnvironVariables();
+
+                if (telnetCommand.size() == 7) { // If no variables are sent, send everything.
+                    qDebug() << "Server requests all NEW_ENVIRON (MNES) variables";
+                        
+                    std::string output;
+                    output += TN_IAC;
+                    output += TN_SB;
+                    output += OPT_NEW_ENVIRON;
+
+                    for (auto it = environVariables.begin(); it != environVariables.end(); ++it) {
+                        output += MNES_IS;   
+                        output += MNES_VAR;
+                        output += it.key().toUtf8().constData();
+                        output += MNES_VAL;
```

Comment:
As `NES_VALUE` the only variables that are supposed to be sent under that type that are likely to be relevant to Mudlet are:
* USER - This variable is used to transmit the user or account name that the client wishes to log into on the remote system.  The format of the value the USER variable is system dependent, as determined by the remote system.
* SYSTEMTYPE - This is used to transmit the type of operating system on the system that sends this variable.  It value is identical to the value of the SYSTEM (SYST) command in [FTP (RFC959)](https://www.rfc-editor.org/rfc/rfc959.txt).  The format of the value shall have as its first word one of the system names listed in the current version of the Assigned Numbers document (See page 86 in https://www.rfc-editor.org/rfc/rfc1340.txt it includes entries for `MACOS` and `WIN32` but no mention {as it is so long out of date} of WIN64, LINUX or FREEBSD for instance).
* DISPLAY - This variable is used to transmit the X display location of the client.  The format for the value of the DISPLAY variable is:
    \<host>:<dispnum>\[.\<screennum>]
    This information is identical to the information passed using the Telnet X-DISPLAY-LOCATION option.  If both the DISPLAY environment variable, and the X-DISPLAY-LOCATION [option RFC1096](https://www.rfc-editor.org/rfc/rfc1096.txt) are received, and they contain conflicting information, the most recently received information received should be used.

The remaining ones being: `JOB`, `ACCT`, `PRINTER` - all other variables should be sent with `NES_USERVAL` ...

Reply:
Implemented USER and SYSTEMTYPE. DISPLAY may be valuable to someone in a future PR. We'll compromise with MNES on NES_VALUE here where we must, but if there is a way to segment into a pure NEW_ENVIRON, I'll try to find it and adjust accordingly. We'll be on the hunt for more candidates for NES_USERVAL.

Reply:
Well the ones that Scandum suggests are required for MNES (according to https://www.tintin.mudhalla.net/protocols/mnes/ ) are:
* "CHARSET" - The CHARSET value could be one of ASCII, BIG5, CP437, CP949, CP1251, EUC-KR, GB18030, ISO-8859-1, ISO-8859-2, KOI8-R, UTF-8. It's generally not feasible for a MUD server to force the client to use a specific character set. The MNES CHARSET value should reflect the current character set mode of the client. See [man charsets](https://linux.die.net/man/7/charsets) for reference. *Though we support an awful lot more encodings that TinTin++ does.* :grinning: 
* "CLIENT_NAME" - Name of the client. *Ours will be `MUDLET` all the time I guess.*
* "CLIENT_VERSION" - Version of the client - the CLIENT_NAME and CLIENT_VERSION variables shouldn't require additional explanation.
* "IPADDRESS" - The client's IP address; it's suggested for proxies adding MTTS support to also report the client's real IP address. This allows a MUD to ban specific users without having to ban an entire proxy. *I see that there have been some discussions already about this!*
* "MTTS" - bitvector. The [MTTS](https://www.tintin.mudhalla.net/protocols/mtts) bitvector that is set by the client should reflect the current MTTS state. See the MTTS specification for more information. *Again you've already done this one.*
* "TERMINAL_TYPE" - name. The terminal type that is returned by the client should reflect the terminal type as described in the [MTTS](https://www.tintin.mudhalla.net/protocols/mtts) specification. *I think we've already decided on `ANSI-TRUECOLOR`.*

Of course if we go down the route of using the proper `NES_USERVAL` it is going to confuse Scandum's parser that is expecting, effectively, `NES_VALUE` for them - he'll just have to remember Postel's Law ("Be liberal in what you accept and conservative in what you send out.") and revise that to accept either. :smiling_imp: 

As such we should declare that we support the **New Environment protocol** option and see how many people moan at Scandum that his *Mud New Environment* doesn't seem to work with Mudlet.